odoo/sql_db.py

```diff
@@ -413,13 +413,26 @@ def __exit__(self, exc_type, exc_value, traceback):
     def savepoint(self):
         """context manager entering in a new savepoint"""
         name = uuid.uuid1().hex
+        env = None
+        for environment in Environment.envs:
+            if environment.cr is self:
+                env = environment
+                break
+        if env:
+            env['base'].flush()
         self.execute('SAVEPOINT "%s"' % name)
         try:
-            yield
+            if env:
+                with env.clear_upon_failure():
+                    yield
+            else:
+                yield
         except Exception:
             self.execute('ROLLBACK TO SAVEPOINT "%s"' % name)
             raise
         else:
+            if env:
+                env['base'].flush()
```

Comment:
The good practice would be to flush when inside a savepoint.
But should it be explicit or implicit (as you do here) ?